### PR TITLE
Add Object Id Support for Managed Identity.

### PR DESCRIPTION
**Pull request by g2vinay**
Closes https://github.com/Azure/azure-sdk-for-java-pr/issues/1424

1. Bumps to latest msal.
2. Adds support to let user specify object Id when using Managed Identity credential.

### CODE REVIEW

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java [hunk 1]
```diff
@@ -172,6 +174,22 @@ public DefaultAzureCredentialBuilder managedIdentityResourceId(String resourceId
...
+     *
+     * Only one of managedIdentityResourceId and managedIdentityClientId can be specified.
+     *
+     * @param objectId the object ID
+     * @return the DefaultAzureCredentialBuilder itself
+     */
+    public DefaultAzureCredentialBuilder managedIdentityObjectId(String objectId) {
+        this.managedIdentityObjectId = objectId;
+        return this;
+    }
```
Comment (billwert) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java:191]: I think we decided to only put this on `ManagedIdentityCredential` - check with @ahsonkhan

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/ValidationUtil.java [hunk 2]
```diff
@@ -156,4 +156,22 @@ public static Path validateSecretFile(File file, ClientLogger logger) {
...
+        }
+        if (resourceId != null) {
+            nonNullIdCount++;
+        }
+        if (objectId != null) {
+            nonNullIdCount++;
+        }
+
+        if (nonNullIdCount > 1) {
+            throw logger.logExceptionAsError(new IllegalStateException("Only one of clientId, resourceId and objectId can be specified."));
```
Comment (billwert) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/util/ValidationUtil.java:174]:
```suggestion
            throw logger.logExceptionAsError(new IllegalStateException("Only one of clientId, resourceId, or objectId can be specified."));
```

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/AzureApplicationCredentialBuilder.java [hunk 3]
```diff
@@ -67,6 +69,22 @@ public AzureApplicationCredentialBuilder managedIdentityResourceId(String resour
...
+     *
+     * Only one of managedIdentityResourceId and managedIdentityClientId can be specified.
+     *
+     * @param objectId the object ID
+     * @return the DefaultAzureCredentialBuilder itself
+     */
+    public AzureApplicationCredentialBuilder managedIdentityObjectId(String objectId) {
+        this.managedIdentityObjectId = objectId;
+        return this;
+    }
```
Comment (billwert) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/AzureApplicationCredentialBuilder.java:86]: I kind of suspect this is the same bucket as the `DefaultAzureCredential` - I think our idea was that if you think you need `objectId` support you are in a very narrow case and you just make a `ManagedIdentityCredential` directly.

**Review by billwert — CHANGES_REQUESTED** [review]
See comments; I think we're putting the new property in places we don't want.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java (hunk 1 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/DefaultAzureCredentialBuilder.java]: scoped to MI cred only.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/AzureApplicationCredentialBuilder.java (hunk 3 above)
Comment (g2vinay) [reply sdk/identity/azure-identity/src/main/java/com/azure/identity/AzureApplicationCredentialBuilder.java]: scoped to MI cred only.

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/ManagedIdentityCredentialBuilder.java [hunk 4]
```diff
@@ -92,18 +94,28 @@ public ManagedIdentityCredentialBuilder resourceId(String resourceId) {
         return this;
     }
 
+    /**
+     * Specifies the object ID of a user assigned or system assigned identity.
+     *
+     * Only one of clientId, resourceId and objectId can be specified.
```
Comment (scottaddie) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/ManagedIdentityCredentialBuilder.java:100]:
```suggestion
     * Specifies the object ID of a user-assigned or system-assigned managed identity.
     *
     * Only one of clientId, resourceId, or objectId can be specified.
```

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 5]
```diff
@@ -1114,6 +1115,12 @@ private Mono<AccessToken> authenticateToServiceFabricManagedIdentityEndpoint(Str
                 payload.append(urlEncode(resourceId));
             }
 
+            if (objectId != null) {
+                LOGGER.warning("User assigned managed identities are not supported in the Service Fabric environment.");
```
Comment (scottaddie) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:1119]:
```suggestion
                LOGGER.warning("User-assigned managed identities are not supported in the Service Fabric environment.");
```

File: sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java [hunk 6]
```diff
@@ -1201,6 +1208,16 @@ public Mono<AccessToken> authenticateToManagedIdentityEndpoint(String identityEn
                 payload.append("&mi_res_id=");
                 payload.append(urlEncode(resourceId));
             }
+
+            if (objectId != null) {
+                if (endpointVersion.equals(MSI_ENDPOINT_VERSION) && headerValue == null) {
+                    // This is the Cloud Shell case. If a clientId is specified, warn the user.
+                    LOGGER.warning("User assigned managed identities are not supported in the Cloud Shell environment.");
```
Comment (scottaddie) [inline sdk/identity/azure-identity/src/main/java/com/azure/identity/implementation/IdentityClient.java:1215]:
```suggestion
                    LOGGER.warning("User-assigned managed identities are not supported in the Cloud Shell environment.");
```

**Review by billwert — APPROVED** [review]
lgtm!

File: eng/versioning/external_dependencies.txt [hunk 7]
```diff
@@ -202,7 +202,7 @@ com.microsoft.azure:azure-mgmt-resources;1.3.0
 com.microsoft.azure:azure-mgmt-search;1.24.1
 com.microsoft.azure:azure-mgmt-storage;1.3.0
 com.microsoft.azure:azure-storage;8.0.0
-com.microsoft.azure:msal4j;1.16.2
+com.microsoft.azure:msal4j;1.17.0
```
Comment (billwert) [inline eng/versioning/external_dependencies.txt:205]: you should rebase onto main; I pulled in 1.17.1 already.

**Review by billwert — APPROVED** [review]

File: sdk/identity/azure-identity/src/test/java/com/azure/identity/ManagedIdentityCredentialTest.java [hunk 8]
```diff
@@ -105,11 +105,24 @@ public void testArcUserAssigned() {
...
+        String objectId = "2323-sd2323s-32323-32334-34343";
 
         // test
         Assertions.assertThrows(IllegalStateException.class,
             () -> new ManagedIdentityCredentialBuilder()
                 .clientId(CLIENT_ID).resourceId(resourceId).build());
+
+        Assertions.assertThrows(IllegalStateException.class,
+            () -> new ManagedIdentityCredentialBuilder()
+                .clientId(CLIENT_ID).resourceId(resourceId).objectId(objectId).build());
```
Comment (billwert) [inline sdk/identity/azure-identity/src/test/java/com/azure/identity/ManagedIdentityCredentialTest.java:117]: do we need this assert? I think the pair testing of the combos with the other assertions covers it?

**Pull request merged**